Sources/SwiftParser/Names.swift

```diff
@@ -11,40 +11,373 @@
     }
     if let dollarIdent = self.consume(if: .dollarIdentifier) {
       return (
-        RawUnexpectedNodesSyntax([dollarIdent], arena: self.arena),
+        RawUnexpectedNodesSyntax(combining: unexpectedSelector, dollarIdent, arena: self.arena),
         self.missingToken(.identifier)
       )
     } else {
       if let wildcardToken = self.consume(if: .wildcard) {
-        return (nil, wildcardToken)
+        return (unexpectedSelector, wildcardToken)
       }
-      return (nil, self.consumeAnyToken(remapping: .identifier))
+      return (unexpectedSelector, self.consumeAnyToken(remapping: .identifier))
     }
   }
 }
 
+private enum StructuralTokens: TokenSpecSet {
+  case comma
+  case colon
+  case leftParen
+  case leftBrace
+  case leftSquare
+  case leftAngle
+  case rightParen
+  case rightBrace
+  case rightSquare
+  case rightAngle
+
+  init?(lexeme: Lexer.Lexeme, experimentalFeatures: Parser.ExperimentalFeatures) {
+    self.init(lexeme.rawTokenKind)
+  }
+
+  init?(_ rawTokenKind: RawTokenKind) {
+    switch rawTokenKind {
+    case .comma: self = .comma
+    case .colon: self = .colon
+    case .leftParen: self = .leftParen
+    case .leftBrace: self = .leftBrace
+    case .leftSquare: self = .leftSquare
+    case .leftAngle: self = .leftAngle
+    case .rightParen: self = .rightParen
+    case .rightBrace: self = .rightBrace
+    case .rightSquare: self = .rightSquare
+    case .rightAngle: self = .rightAngle
+    default: return nil
+    }
+  }
+
+  var spec: TokenSpec {
+    switch self {
+    case .comma: return .comma
+    case .colon: return .colon
+    case .leftParen: return .leftParen
+    case .leftBrace: return .leftBrace
+    case .leftSquare: return .leftSquare
+    case .leftAngle: return .leftAngle
+    case .rightParen: return .rightParen
+    case .rightBrace: return .rightBrace
+    case .rightSquare: return .rightSquare
+    case .rightAngle: return .rightAngle
+    }
+  }
+}
+
+extension TokenConsumer {
+  /// Do the subsequent tokens have the form of a module selector? Encompasses some invalid syntax which nonetheless
+  /// can be handled by `consumeModuleSelectorTokens()`.
+  ///
+  /// - Postcondition: If `true`, either the current token or the next token is `.colonColon`.
+  mutating func isAtModuleSelector() -> Bool {
+    // If this is a module selector, the next token should be `::`.
+    guard self.peek(isAt: .colonColon) else {
+      // ...however, we will also allow the *current* token to be `::`. `consumeModuleSelectorTokens()` will create a
+      // missing identifier.
+      return self.at(.colonColon)
+    }
+
+    // Technically the current token *should* be an identifier, but we also want to diagnose other tokens that might be
+    // used by accident (particularly keywords and `_`). However, we don't want to consume tokens which would make the
+    // surrounding structure mis-parse.
+    return self.at(anyIn: StructuralTokens.self) == nil
```

Comment:
I don't think it's a good idea to always consider `<token>` `::` as a module qualifier. E..g.
```
class
::
```
This looks to me an incomplete `class` declaration with just an orphan `::`. Even if it's on the same line with no-space, I don't think we need to parse it as a module selector.